### PR TITLE
#144 - Added Storage.Wrap a forwarding decorator for Storage interface

### LINKED ISSUE #144: Add abstract decorator for Storage
It is inconvenient to create implementations of Storage composed of other implementations without an abstract decorator class

### PR DESCRIPTION
**Pull request by olegmoz**
Closes #144 
Added Storage.Wrap - abstract decorator for Storage

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#145](https://github.com/artipie/asto/issues/145) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/63779ce9-5d3b-489e-91f8-3539a2e096cc); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/asto/Storage.java [hunk 1]
```diff
@@ -101,4 +101,61 @@
      * @return Transaction
      */
     CompletableFuture<Transaction> transaction(List<Key> keys);
+
+    /**
+     * Forwarding decorator for {@link Storage}.
+     *
+     * @since 0.17
```
Comment (Vatavuk) [inline src/main/java/com/artipie/asto/Storage.java:108]: @olegmoz since 0.18

File: src/main/java/com/artipie/asto/Storage.java [hunk 2]
```diff
@@ -101,4 +101,61 @@
...
+         * Ctor.
+         *
+         * @param delegate Delegate storage
+         */
+        protected Wrap(final Storage delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public CompletableFuture<Boolean> exists(final Key key) {
```
Comment (Vatavuk) [inline src/main/java/com/artipie/asto/Storage.java:127]: @olegmoz all methods should be final

**Comment by olegmoz** [comment]
@Vatavuk thanks for the review, everything has been fixed. Please review

**Comment by 0crat** [comment]
Job was finished in 21 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@Vatavuk please make sure you will find at least 3 issues during next CR, thanks

**Comment by Vatavuk** [comment]
@sereshqua ok

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Pull request merged**